panels/user-accounts/cc-avatar-chooser.c

```diff
@@ -410,35 +410,61 @@ cheese_camera_device_monitor_new_cb (GObject *source,
 }
 #endif /* HAVE_CHEESE */
 
-static void
-setup_photo_popup (CcAvatarChooser *self)
+static GSList *
+get_settings_facesdirs (void)
 {
-        GFile *file, *dir;
-        GFileInfo *info;
-        GFileEnumerator *enumerator;
-        GFileType type;
-        const gchar *target;
-        const gchar * const * dirs;
-        guint i;
-        gboolean added_faces;
+        GSList *facesdirs = NULL;
+        char **settings_dirs;
+        int i;
 
-        self->faces = g_list_store_new (G_TYPE_FILE);
-        gtk_flow_box_bind_model (GTK_FLOW_BOX (self->flowbox),
-                                 G_LIST_MODEL (self->faces),
-                                 create_face_widget,
-                                 self,
-                                 NULL);
+        g_autoptr(GSettings) settings = NULL;
 
-        g_signal_connect (self->flowbox, "child-activated",
-                          G_CALLBACK (face_widget_activated), self);
+        settings = g_settings_new_with_path ("org.gnome.controlcenter.user-accounts",
+                                             "/org/gnome/controlcenter/user-accounts/");
 
-        dirs = g_get_system_data_dirs ();
-        for (i = 0; dirs[i] != NULL; i++) {
-                char *path;
+        settings_dirs = g_settings_get_strv (settings, "facesdirs");
+        if (settings_dirs != NULL) {
+                for (i = 0; settings_dirs[i] != NULL; i++) {
+                        char *path = settings_dirs[i];
+                        if (path != NULL && g_strcmp0 (path, "") != 0)
+                                facesdirs = g_slist_prepend (facesdirs, g_strdup (path));
+                }
+                g_strfreev (settings_dirs);
+        }
 
-                path = g_build_filename (dirs[i], "pixmaps", "faces", NULL);
+        return g_slist_reverse (facesdirs);
+}
+
+static GSList *
+get_system_facesdirs (void)
+{
+        GSList *facesdirs = NULL;
+        const char * const * data_dirs;
+        int i;
+
+        data_dirs = g_get_system_data_dirs ();
+        for (i = 0; data_dirs[i] != NULL; i++) {
+                char *path = g_build_filename (data_dirs[i], "pixmaps", "faces", NULL);
+                facesdirs = g_slist_prepend (facesdirs, path);
+        }
+
+        return g_slist_reverse (facesdirs);
+}
+
+static gboolean
+add_faces_from_dirs (GListStore *faces, GSList *facesdirs, gboolean add_all)
```

Comment:
Why make the `add_all` distinction between system and settings faces dirs? It seems unexpected to only pick faces from the first dir specified in the settings, but all of the system dirs, but I could be convinced otherwise if you have a reason.

Reply:
Because that's the current behavior in gnome-control-center. If there are more than one directory in the `g_get_system_data_dirs`, we'll have more than one folder but we only want to show the first one with valid faces.

In the case of a setting, we'll show faces from all directories in that setting, that's what I was thinking about having this as a list of directories. The other option could be to do the same as the system, just find the first directory with valid faces, but don't combine all. The second case is easier because we can get all directories in one list and use the first with real images.

Reply:
OK, got it. Then I agree it makes sense to preserve the current behaviour from the system directories, and combine the settings directories.